s3/s3.cc

```diff
@@ -0,0 +1,625 @@
+    }
+
+    auto response_code = std::atoi(status_line.substr(i, e - i).begin());
+
+    size_t content_length = 0;
+
+    // Parse response headers
+    std::unordered_map<sstring, sstring> response_headers;
+    while (1) {
+        co_await in.consume(lc);
+        auto line = lc.get();
+        if (line.empty()) {
+            break;
+        }
+        log.debug("conn {}: response header: {}", fmt::ptr(&*con), line);
+        auto hdr = parse_header(line);
+        if (hdr.first == "Content-Length") {
+            content_length = std::atoi(hdr.second.begin());
+        }
+        response_headers.emplace(std::move(hdr));
+    }
+
+    temporary_buffer<char> response_content;
+    if (response_code != 204) { // 204 = No Content
+        response_content = co_await in.read_exactly(content_length);
+    }
+
+    if (response_code < 200 || response_code >= 300) {
+        throw std::runtime_error(format("request failed: {}", status_line));
+    }
+
+    co_return http_response{response_code, content_length, std::move(response_content), std::move(response_headers)};
+}
+
+// Keep headers alive around async operation.
+static
+future<http_response> make_http_request(connection_ptr con,
+                                        const sstring& method,
+                                        const sstring& path,
+                                        const std::map<sstring, sstring>& headers = {},
+                                        std::optional<sstring> content = std::nullopt) {
+    if (!content) {
+        std::optional<http_content_writer> wr = std::nullopt;
+        return make_http_request(con, method, path, headers, std::move(wr));
+    }
+    auto content_size = content->size();
+    return make_http_request(con, method, path, headers, http_content_writer{content_size,
+         [content = std::move(content)] (output_stream<char>& out) {
+             return out.write(*content);
+         }}
+    );
+}
+
+sstring parse_xml_tag(std::string_view xml, const sstring& name) {
+    auto i = xml.find(name);
+    if (i == std::string_view::npos) {
+        throw_with_backtrace<std::runtime_error>(format("Response does not contain <{}>: {}", name, xml));
+    }
+    i += name.size();
+    if (i == xml.size() || xml[i] != '>') {
+        throw_with_backtrace<std::runtime_error>(format("Response does not contain <{}>: {}", name, xml));
+    }
+    ++i;
+    auto e = xml.find(name, i);
+    if (e == std::string_view::npos) {
+        throw_with_backtrace<std::runtime_error>(format("Response does not contain </{}>: {}", name, xml));
+    }
+    if (e == i || xml[e - 1] != '/') {
+        throw_with_backtrace<std::runtime_error>(format("Response does not contain </{}>: {}", name, xml));
+    }
+    --e;
+    if (e == i || xml[e - 1] != '<') {
+        throw_with_backtrace<std::runtime_error>(format("Response does not contain </{}>: {}", name, xml));
+    }
+    --e;
+    return sstring(xml.begin() + i, xml.begin() + e);
+}
```

Comment:
A local http is understandable, but XML? 

A random search suggested pugixml (DOM) or expat (events).